src/Compilers/CSharp/Portable/Symbols/Source/AwaitInAsyncMethodChecker.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed class AwaitInAsyncMethodChecker : CSharpSyntaxWalker
+    {
+        internal static void Check(SyntaxNode syntax, Location location, DiagnosticBag diagnostics)
+        {
+            var awaitFinder = new AwaitInAsyncMethodChecker();
+
+            switch (syntax.Kind())
+            {
+                case SyntaxKind.ParenthesizedLambdaExpression:
+                    syntax = ((ParenthesizedLambdaExpressionSyntax)syntax).Body;
+                    break;
+                case SyntaxKind.SimpleLambdaExpression:
+                    syntax = ((SimpleLambdaExpressionSyntax)syntax).Body;
+                    break;
+                case SyntaxKind.LocalFunctionStatement:
+                    var local = (LocalFunctionStatementSyntax)syntax;
+                    syntax = (SyntaxNode)local.Body ?? local.ExpressionBody;
+                    break;
+            }
```

Comment:
Does each caller already have access to the appropriately typed syntax? If so, perhaps the caller could pass in the `Body` directly.

Reply:
Some do (Method and Local Function), but some don't (Lambda).